### PR TITLE
Bigc 637 api cart delete simple product

### PR DESCRIPTION
**Pull request by aaron-martin**
@sofib 
Please let me know if you are not able to do this review. Sorry to stick you with another. 

Thanks
Aaron

### CODE REVIEW

File: extension/lib/bigcommerce/CartRepository.js [hunk 1]
```diff
@@ -138,6 +138,23 @@ class BigCommerceCartRepository {
 
     return response.data.checkout_url
   }
+
+  /**
+   * @param {Array} cartItemIds
```
Comment (sofib) [inline extension/lib/bigcommerce/CartRepository.js]: This should rather be an array of strings.

File: extension/lib/shopgate/read_model/AddProductsToCart.js [hunk 2]
```diff
@@ -1,3 +0,0 @@
-/**
```
Comment (sofib) [inline extension/lib/shopgate/read_model/AddProductsToCart.js]: What is the reason for deleting this typedef?

File: extension/lib/deleteProductFromCart.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
+const ShopgateCartPipeline = require('./shopgate/CartExtensionPipeline')
+
+/**
+ * @param {PipelineContext} context
+ * @param {AddProductsInput} input
+ * @returns {Promise<void>}
+ */
+module.exports = async (context, input) => {
+  try {
+    await ShopgateCartPipeline.create(context).deleteProductFromCart(input.CartItemIds)
```
Comment (sofib) [inline extension/lib/deleteProductFromCart.js]: `input` is not documented.

File: extension/lib/shopgate/CartExtensionPipeline.js [hunk 4]
```diff
@@ -86,6 +86,14 @@ class ShopgateCartExtensionPipeline {
     return this._bigCommerceCartRepository.getCheckoutUrl()
   }
 
+  /**
+   * @param {Array} cartItemIds
```
Comment (sofib) [inline extension/lib/shopgate/CartExtensionPipeline.js]: This should rather be an array of strings.

File: extension/lib/shopgate/CartExtensionPipeline.js [hunk 5]
```diff
@@ -86,6 +86,14 @@ class ShopgateCartExtensionPipeline {
...
+   * @return {Promise.<void>}
```
Comment (sofib) [inline extension/lib/shopgate/CartExtensionPipeline.js]: We're usually using syntax without dot: `Promise<void>`.

File: test/unit/extension/lib/bigcommerce/CartRepositoryTest.js [hunk 6]
```diff
@@ -106,4 +106,28 @@ describe('BigCommerceCartRepository - unit', function () {
 
     return subjectUnderTest.getCheckoutUrl().should.eventually.be.rejectedWith(Error)
   })
+
+  it('should properly call bigCommerce client delete method when doing a deleteProductsFromCart', function () {
```
Comment (sofib) [inline test/unit/extension/lib/bigcommerce/CartRepositoryTest.js]: Maybe we could make a test that is able to show that every item id passed inside of array will be deleted, not just one.

File: test/unit/extension/lib/bigcommerce/CartRepositoryTest.js [hunk 7]
```diff
@@ -106,4 +106,28 @@ describe('BigCommerceCartRepository - unit', function () {
...
+    const cartItemIds = ['abc-def-ghi-jk']
+    storageMock.expects('get').once().returns('0000-0000-0000-0000')
+    bigCommerceMock.expects('delete').withArgs('/carts/0000-0000-0000-0000/items/' + cartItemIds[0])
+
+    return subjectUnderTest.deleteProductFromCart(cartItemIds).should.eventually.be.fulfilled
+  })
+
+  it('should see error when bigCommerce delete method throws and error', function () {
```
Comment (sofib) [inline test/unit/extension/lib/bigcommerce/CartRepositoryTest.js]: Maybe try forming sentencies according to https://dzone.com/articles/7-popular-unit-test-naming number five:
*Should_ExpectedBehavior_When_StateUnderTest*
inoour case to make a difference between other tests methods as well we'd need something like: should fail deleting cart items when bigcommerce delete operation fails

File: test/unit/extension/lib/bigcommerce/CartRepositoryTest.js [hunk 8]
```diff
@@ -106,4 +106,28 @@ describe('BigCommerceCartRepository - unit', function () {
...
+    const cartItemIds = ['abc-def-ghi-jk']
+    storageMock.expects('get').once().returns('0000-0000-0000-0000')
+    bigCommerceMock.expects('delete').withArgs('/carts/0000-0000-0000-0000/items/' + cartItemIds[0]).throws(Error)
+
+    return subjectUnderTest.deleteProductFromCart(cartItemIds).should.eventually.be.rejectedWith(Error)
+  })
+
+  it('should see error when cartId is not returned during delete', function () {
```
Comment (sofib) [inline test/unit/extension/lib/bigcommerce/CartRepositoryTest.js]: See previous comment for test naming .

File: extension/lib/deleteProductFromCart.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
+const ShopgateCartPipeline = require('./shopgate/CartExtensionPipeline')
+
+/**
+ * @param {PipelineContext} context
+ * @param {AddProductsInput} input
```
[the remark this answers is not in the document]
Comment (aaron-martin) [reply extension/lib/deleteProductFromCart.js]: corrected this bit of JSDoc.

File: pipelines/deleteProductsFromCart_v1.json [hunk 10]
```diff
@@ -0,0 +1,51 @@
+{
+  "version": "1",
+  "pipeline": {
+    "id": "deleteProductsFromCart_v1",
+    "input": [
+      {
+        "id": "1010",
```
[the remark this answers is not in the document]
Comment (aaron-martin) [reply pipelines/deleteProductsFromCart_v1.json]: removed sgxsMeta input from pipeline input

File: pipelines/deleteProductsFromCart_v1.json [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+      }
+    ],
+    "steps": [
+      {
+        "id": "@shopgate/bigcommerce-cart",
+        "path": "@shopgate/bigcommerce-cart/lib/deleteProductFromCart.js",
+        "type": "extension",
+        "input": [
+          {
+            "id": "1010",
```
[the remark this answers is not in the document]
Comment (aaron-martin) [reply pipelines/deleteProductsFromCart_v1.json]: removed sgxsMeta input from extension step

File: extension/lib/bigcommerce/CartRepository.js (hunk 1 above)
Comment (aaron-martin) [reply extension/lib/bigcommerce/CartRepository.js]: done

File: extension/lib/shopgate/CartExtensionPipeline.js (hunk 4 above)
Comment (aaron-martin) [reply extension/lib/shopgate/CartExtensionPipeline.js]: done

File: extension/lib/shopgate/CartExtensionPipeline.js (hunk 5 above)
Comment (aaron-martin) [reply extension/lib/shopgate/CartExtensionPipeline.js]: done

File: test/unit/extension/lib/bigcommerce/CartRepositoryTest.js (hunk 7 above)
Comment (aaron-martin) [reply test/unit/extension/lib/bigcommerce/CartRepositoryTest.js]: Changed name to hopefully better conform the test naming conventions

File: test/unit/extension/lib/bigcommerce/CartRepositoryTest.js (hunk 8 above)
Comment (aaron-martin) [reply test/unit/extension/lib/bigcommerce/CartRepositoryTest.js]: Changed name to hopefully better conform the test naming conventions

File: test/unit/extension/lib/bigcommerce/CartRepositoryTest.js (hunk 6 above)
Comment (aaron-martin) [reply test/unit/extension/lib/bigcommerce/CartRepositoryTest.js]: I modified test that tests successful delete from cart so it deletes two products. Utilized onCall() to check that the bigCommerce client delete method is called twice with the correct parameters based on the item being deleted.

File: extension/lib/deleteProductFromCart.js [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+ * @returns {Promise<void>}
```
[the remark this answers is not in the document]
Comment (aaron-martin) [reply extension/lib/deleteProductFromCart.js]: I tested and returning an empty object was not needed. I removed it.

File: extension/lib/shopgate/read_model/Shopgate.js [hunk 13]
```diff
@@ -44,3 +44,8 @@
  * @property {string} storeHash
  * @property {string} clientSecret
  */
+
+/**
+ * @typedef {Object} DeleteProductsInput
+ * @property {[string]} CartItemIds
```
Comment (sofib) [inline extension/lib/shopgate/read_model/Shopgate.js]: The way we document arrays is for example `string[]`.

File: extension/lib/shopgate/CartExtensionPipeline.js [hunk 14]
```diff
@@ -86,6 +86,14 @@ class ShopgateCartExtensionPipeline {
     return this._bigCommerceCartRepository.getCheckoutUrl()
   }
 
+  /**
+   * @param {[string]} cartItemIds
```
Comment (sofib) [inline extension/lib/shopgate/CartExtensionPipeline.js]: The way we document arrays is for example `string[]`.

File: extension/lib/deleteProductFromCart.js (hunk 3 above)
Comment (sofib) [reply extension/lib/deleteProductFromCart.js]: This is resolved.

File: extension/lib/shopgate/read_model/AddProductsToCart.js (hunk 2 above)
Comment (sofib) [reply extension/lib/shopgate/read_model/AddProductsToCart.js]: What about this question?

Comment (aaron-martin) [reply extension/lib/shopgate/read_model/AddProductsToCart.js]: I re-added this file.

File: extension/lib/shopgate/CartExtensionPipeline.js (hunk 14 above)
Comment (aaron-martin) [reply extension/lib/shopgate/CartExtensionPipeline.js]: corrected

File: extension/lib/shopgate/read_model/Shopgate.js (hunk 13 above)
Comment (aaron-martin) [reply extension/lib/shopgate/read_model/Shopgate.js]: corrected

**Pull request merged**